src/Domain/Python/BlockLogicalCoordinates.cpp

```diff
@@ -0,0 +1,75 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/Python/BlockLogicalCoordinates.hpp"
+
+#include <array>
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/IdPair.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/BlockLogicalCoordinates.hpp"
+#include "Domain/Domain.hpp"
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
+#include "Domain/Structure/BlockId.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace py = pybind11;
+
+namespace domain::py_bindings {
+
+namespace {
+template <size_t Dim>
+void bind_block_logical_coordinates_impl(py::module& m) {  // NOLINT
+  py::class_<
+      IdPair<domain::BlockId, tnsr::I<double, Dim, Frame::BlockLogical>>>(
+      m, ("BlockIdAndLogicalCoord" + get_output(Dim) + "D").c_str());
+  m.def(
+      "block_logical_coordinates",
+      [](const Domain<Dim>& domain,
+         const std::array<DataVector, Dim>& inertial_coords,
+         const std::optional<double>& time = std::nullopt,
+         const std::optional<std::unordered_map<
+             std::string, const domain::FunctionsOfTime::FunctionOfTime&>>&
+             functions_of_time = std::nullopt) {
+        const tnsr::I<DataVector, Dim, Frame::Inertial> x{inertial_coords};
+        // Transform functions-of-time map to unique_ptrs because pybind11 can't
+        // handle unique_ptrs easily as function arguments (it's hard to
+        // transfer ownership of a Python object to C++)
+        std::unordered_map<
+            std::string,
+            std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+            functions_of_time_ptrs{};
+        if (functions_of_time.has_value()) {
+          for (const auto& [name, fot] : *functions_of_time) {
```

Comment:
can this call our `clone_unique_ptrs` function instead? I think that does the same thing. You might even be able to do:
```
block_log... (..., clone_unique_ptrs(functions_of_time.value_or(unordered_map...{}));
```

Reply:
The argument type here is a map of const-refs, not unique_ptrs, so I think clone_unique_ptrs doesn't have a suitable overload. (const-refs can be passed easily from python, unique_ptrs are harder to pass because apparently it's hard to transfer ownership of an object from Python to C++)

Reply:
Oh right... Sorry!